k256/Cargo.toml

```diff
@@ -20,6 +20,7 @@ rust-version = "1.60"
 [dependencies]
 cfg-if = "1.0"
 elliptic-curve = { version = "0.12.3", default-features = false, features = ["hazmat", "sec1"] }
+once_cell = { version = "1.16", default-features = false, features = ["critical-section"] }
```

Comment:
:thinking: not sure that enabling critical-section _here_ is the right call. I'd imagine it should be up to the leaf crate with Cargo.lock to enable this, such that pulling this crate doesn't pull cs unconditionally. 

One imporovement here would be to add `std = "once-cell/std"`, so that once_cell doesn't use critical section (Cargo would still download and compile this crate, it'll just be unused). 

But I think what you should do is:

```
[dependencies]
once_cell = { version = "1.16", default-features = false }

[features]
default = ["arithmetic", "ecdsa", "pkcs8", "schnorr", "std"]
alloc = ["ecdsa-core?/alloc", "elliptic-curve/alloc"]

# Either std or critical-section should be enabled
std = ["once-cell/std", "alloc", "ecdsa-core?/std", "elliptic-curve/std"]
critical-section = ["once-cell/critical-section"]
```

That way:

* people with `std` don't get `critical_section` at all
* people with `no_std` would have to explicitly opt-into critical section
  * that's a bit bad because it doesn't "just work"
  * OTOH, silently opting the users in critical-section behavior feels not super friendly. It's a big hammer, better to be explicitly aware about it!

Reply:
Sounds good to me. Want to make a PR?

FWIW, the situation is probably not as bad as it seems, because I expect most embedded users will want to disable `precomputed-tables` anyway to reduce the size of the library in memory

Reply:
> Want to make a PR?

Not really :) 

>FWIW, the situation is probably not as bad as it seems

This is a bit ambiguous, but, with the current code, my main worry is std people who get opted into critical-section. I am not sure how exactly critical-section behaves on std, but at minimum that hits much less tested and much slower path of once_cell, _for all once cells in the final binary_, and, at maximum, that actually serializes access to all once cells everywhere :fearful:

Reply:
Heh ok, I'll make a PR

Reply:
>  for all once cells in the final binary

I guess the `once_cell` API has the same problems as `lazy_static`. Big yikes. Hopefully it won't be the case when it's a part of the language proper.

Reply:
std won't have this problem by virtue of not supporting no_std at all. If you want to configure synchronisation for a _specific_ once_cell, rather than for all of them globally, https://crates.io/crates/generic_once_cell is the right crate.

Reply:
Ah, so it's going to be std-only? That's a bummer.

Reply:
@fjarri with an `alloc` dependency it would be possible to use `race::OnceBox`